server/license: Add notice for upcoming throttling

spilchen opened the pull request:
Previously, throttling errors would only occur when active throttling was in effect. This update adds notices to alert clients when throttling is imminent, without disrupting running queries. A notice will be returned in the following cases:
- If no license is installed or the license has expired, but we are still within the grace period.
- If the license requires telemetry and no telemetry data has been received in the past 3 days.

In both cases, a notice is triggered only if the max limit of 5 concurrent open transactions is exceeded.

Additionally, notices are logged to cockroach.log at most once every 5 minutes. This commit also cleans up prior logging code by utilizing log.Every() for more efficient log gating.

This will be backported to 24.2, 24.1, 23.2 and 23.1.

Epic: CRDB-39988
Closes CRDB-39992
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131284)
<!-- Reviewable:end -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131284#-:-O7ZBVNH4Y84Xg0rkr9a:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/server/license/enforcer.go` line 576 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131284#-O7ZARSF5S20DOhy50VR:-O7ZARSF5S20DOhy50VS:ba650aw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e2e5d99a4c81c83b93ccdf9c8a93d102e17d56b/pkg/server/license/enforcer.go#L576)):*
> ```Go
> // getLicenseExpiryTS returns the license expiration timestamp.
> // If no license is installed, it returns false for the second parameter.
> func (e *Enforcer) getLicenseExpiryTS() (time.Time, bool) {
> ```

Nit : Use named return parameters


<!-- Sent from Reviewable.io -->

spilchen committed:
server/license: Add notice for upcoming throttling

Previously, throttling errors would only occur when active throttling was in
effect. This update adds notices to alert clients when throttling is imminent,
without disrupting running queries. A notice will be returned in the following
cases:
- If no license is installed or the license has expired, but we are still
  within the grace period.
- If the license requires telemetry and no telemetry data has been received in
  the past 3 days.

In both cases, a notice is triggered only if the max limit of 5 concurrent open
transactions is exceeded.

Additionally, notices are logged to cockroach.log at most once every 5 minutes.
This commit also cleans up prior logging code by utilizing log.Every() for more
efficient log gating.

This will be backported to 24.2, 24.1, 23.2 and 23.1.

Epic: CRDB-39988
Closes CRDB-39992
Release note: None

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131284)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @fqazi)

___
*[`pkg/server/license/enforcer.go` line 576 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131284#-O7ZARSF5S20DOhy50VR:-O7ZWFzL5iEbLNj0IytV:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e2e5d99a4c81c83b93ccdf9c8a93d102e17d56b/pkg/server/license/enforcer.go#L576)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit : Use named return parameters
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR @fqazi 

bors r=fqazi
